rosflight/include/rosflight/mavrosflight/logger_interface.h

```diff
@@ -0,0 +1,202 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file logger_interface.h
+ * @author Jacob Willis <jbwillis272@gmail.com>
+ */
+
+#ifndef MAVROSFLIGHT_LOGGER_INTERFACE_H
+#define MAVROSFLIGHT_LOGGER_INTERFACE_H
+
+#include <sstream>
+
+#include <stdio.h>
+
+namespace mavrosflight
+{
+/**
+ * \class LoggerInterface
+ * \brief Abstract base class for message handler
+ *
+ * The implementations of this class define how messages are displayed, logged,
+ * etc. To create custom behavior, a derived class should implement each of the
+ * public functions.
+ */
+template <typename Derived>
+class LoggerInterface
+{
+public:
+  template <typename... T>
+  void debug(const char* format, const T&... args)
+  {
+    Derived& derived = static_cast<Derived&>(*this);
+    derived.debug(format, args...);
+  }
+  template <typename... T>
+  void debug_throttle(float period, const char* format, const T&... args)
+  {
+    Derived& derived = static_cast<Derived&>(*this);
+    derived.debug_throttle(period, format, args...);
+  }
+
+  template <typename... T>
+  void info(const char* format, const T&... args)
+  {
+    Derived& derived = static_cast<Derived&>(*this);
+    derived.info(format, args...);
+  }
+  template <typename... T>
+  void info_throttle(float period, const char* format, const T&... args)
+  {
+    Derived& derived = static_cast<Derived&>(*this);
+    derived.info_throttle(period, format, args...);
+  }
+
+  template <typename... T>
+  void warn(const char* format, const T&... args)
+  {
+    Derived& derived = static_cast<Derived&>(*this);
+    derived.warn(format, args...);
+  }
+  template <typename... T>
+  void warn_throttle(float period, const char* format, const T&... args)
+  {
+    Derived& derived = static_cast<Derived&>(*this);
+    derived.warn_throttle(period, format, args...);
+  }
+
+  template <typename... T>
+  void error(const char* format, const T&... args)
+  {
+    Derived& derived = static_cast<Derived&>(*this);
+    derived.error(format, args...);
+  }
+  template <typename... T>
+  void error_throttle(float period, const char* format, const T&... args)
+  {
+    Derived& derived = static_cast<Derived&>(*this);
+    derived.error_throttle(period, format, args...);
+  }
+
+  template <typename... T>
+  void fatal(const char* format, const T&... args)
+  {
+    Derived& derived = static_cast<Derived&>(*this);
+    derived.fatal(format, args...);
+  }
+  template <typename... T>
+  void fatal_throttle(float period, const char* format, const T&... args)
+  {
+    Derived& derived = static_cast<Derived&>(*this);
+    derived.fatal_throttle(period, format, args...);
+  }
+};
+
+/**
+ * \class DefaultLogger
+ * \brief Default logger that outputs to stdout and stderr
+ */
+class DefaultLogger : public LoggerInterface<DefaultLogger>
```

Comment:
I'm wondering if it's worth moving this into its own header file? `rosflight/mavrosflight/default_logger.h`?